### ISSUE TITLE
Dedicated backend for Storj S3

### ISSUE THREAD

**Issue by DarkCaster** [s3ql/s3ql#326]
UPD:
Based on the results of the discussion I'll try to maintain my sub-optimal solution as a separate fork: https://github.com/DarkCaster/s3ql

### The Problem

There is such a service as storj.io. They provide S3 compatible storage at extremely affordable prices.
Unfortunately, their S3 API has one major hidden problem that cause file system failure in a situation when the number of stored objects crosses the 100,000 mark.

It's all about ListObjectsV2 API call, which is used during synchronization of S3QL metadata, and for operations like fsck. The “prefix” parameter works as described in the S3 specification only until the number of objects in the bucket is less than 100,000. When this mark is reached, calling this API method leads to an error that immediately breaks the file system without the ability to even simply repair it and restart it (because this call is also used in fsck). The problem was NOT documented properly in their docs, at least before I contacted support about it.

I spent the evening talking with their development team, and they suggested a possible solution to this problem

### The Solution

ListObjectsV2 used at their S3 gateway actually can work on a buckets with more than 100K objects in it, but with one particular limitation, it can only process prefix that ended with forward slash delimiter.

So, I spend a week to implement solution that use this behavior. The idea is pretty simple: I can rename all s3ql_data_<number> objects to the s3ql_data/<number> form, so we can request data chunks from fsck with "s3ql_data/" prefix instead "s3ql_data_" prefix.
For other object names - I simply put it under "s3ql_other/" prefix, and filter it manually on client side, fortunately, the number of such objects is quite small.

So, I made all required changes as separate backend class: "storjs3". It inherits from s3c backend class, and perform all object names and prefix translation on the fly.

So far it is working with my s3ql deployment with ~2TB of data and 102K of objects, the filesystem may be transformed manually with scripts, even in a broken state, and fsck may be run after that - however storjs3 backend must be set as substitute to s3c backend at prefix_map so local cache and metadata may be used properly while running fsck (you may remove old cache and use separate storjs3 backend after that, it will create proper cache for it). All the object move process done at the server side, so no need to upload/download anything (hovewer the process may cost you traffic equal to storage size).

Thus, I would like to suggest that you include my changes in the main branch of the project. Unfortunately, I'm not a Python programmer. So my coding style may be bad. Also I do not know how to add tests for new backend. Currently it passes all s3c backend tests if I register it as s3c backend at prefix_map.

Also I'm running an outdated version of S3QL from my own fork of your repo: https://github.com/DarkCaster/s3ql , all the changes located at this file: https://github.com/DarkCaster/s3ql/blob/custom/src/s3ql/backends/storjs3.py . Can you please, examine, what needs to be done to include my changes into mainline ? I need some sort of code-review, and, maybe some help to write the tests. I can also provide temporary access to storj bucket from my account for you to run some tests on real backend if needed. Also you may ask for details about badly implemented ListObjectsV2 S3 API from storj dev team, I talked about this with guy named  Dominick Marino. He may know better about their S3 API compatibility issues. I can give him a link to this issue if you want.

Thanks!

**Comment by Nikratio** [comment]
Thanks for the detailed problem description! 

In order for me to take a look at this, could you please rebase your changes onto the current master branch and submit a pull request?

**Comment by DarkCaster** [comment]
Sure! I'll do this.

**Comment by DarkCaster** [comment]
I've synchronized my changes with the latest upstream, and updated storjs3 class to use new readinto_fh/write_fh methods.
Have not tested it with real data yet, but seem it pass the tests if I register it as s3c backend at prefix_map.

**Comment by acolombier** [comment]
Potentially silly question, but could it be possible to make this a configurable behavior on the `s3c` backend instead using something like `backend-option` ? 
My fear being, there is probably other vendors with some flaws in how they implement the S3 API, and it would probably be best not to have a custom backend for each vendors. 
At the minute, there is an AWS specifc implementation tho, so perhaps having a backend for each vendor is the way forward, but I'm not sure how this will be maintainable in the future in terms of testing.

**Comment by DarkCaster** [comment]
The main problem with Storj - the only way to fix the problem with their API is to reorganize the objects layout. So this storage scheme may not me compatible with any other S3C backend. That's why I implemented it as separate backend.

**Comment by acolombier** [comment]
Yes, that makes sense, but isn't a matter of "customizing" the prefix, as well as using various prefixes for each object types?
Another question on that 100K limit, is your workaround working in case there is more than 100k object of the same type (e.g 100k+ `s3ql_data/...`)?

**Comment by DarkCaster** [comment]
> Another question on that 100K limit, is your workaround working in case there is more than 100k object of the same type (e.g 100k+ `s3ql_data/...`)?

Affirmative. Currently I have more than 100K s3ql_data/ objects. ListObjectsV2 calls for my bucket will only work now for prefixes ended with '/' slash character.

**Comment by DarkCaster** [comment]
> but isn't a matter of "customizing" the prefix, as well as using various prefixes for each object types?

I tried such behavior. You may see it in some early commits in my pull request. Code is just too bulky. For my 2Tb bucket I only have 45 other non s3ql_data objects. It's just easier to filter it manually when we need to list them.

**Comment by Nikratio** [comment]
Thanks for submitting the pull request! 

On the plus side, this is a nicely self contained solution that is unlikely to cause problems for anyone else. On the other side, having the backend do transformations on the key name based on assumptions about which `prefix` parameters the `list` object may receive in the future is very fragile and hacky.

Before we make a decision on this, let's look into other possible options. Firstly, have you checked how and where S3QL actually uses `list` with `prefix` parameter? I could imagine that we could also work around the issue there.

**Comment by DarkCaster** [comment]
If I understood it right - In normal everyday FS use, backend.list mainly used during metadata synchronization. For other cases - it actively used in various helper utilities (and that totally make sense because it purpose to fix/convert/create filesystem so it needed to query bucket for objects of various types/prefixes).

There also much more possible indirect uses of backend.list via iterators and other internal stuff in AbstractBackend like __contains__ , iteritems , etc.

I want to remind that I'm still using old s3ql 4.0 in production. So my findings and assumptions may be not correct for the latest FS version. I'll try to upgrade my FS to the latest version soon and test it (when I'm finished with backups after that massive FS crash that was the reason why I made this new backend)

I'm not sure if it even possible get rid of prefix and key name assumptions without fully rewriting the whole FS. Basically all we need is to classify object designations somehow - is it data object or metadata object. And treat it differently based on this. So, currently I'm classifying objects by prefix (and only assuming we have 2 types of objects - data objects and all others). Just reminding - I'm not python developer and I'm not sure I can correctly add an alternative way to classify objects. Anyway this will me a massive change compared to proposed classification based prefix name.

**Comment by Nikratio** [comment]
Sorry for not being clear. I didn't mean looking at what S3QL is doing at runtime, but looking at the code that calls `backend.list`. Here's what I see:

```
# ack backend.list src/
src/s3ql/database.py
472:    for obj in backend.list('s3ql_metadata_'):
563:    for obj in backend.list('s3ql_params_'):

src/s3ql/backends/comprenc.py
302:        return self.backend.list(prefix)

src/s3ql/fsck.py
1044:            for i, obj_name in enumerate(self.backend.list('s3ql_data_')):

src/s3ql/adm.py
248:    for i, obj_id in enumerate(backend.list()):
```

If I understood you correctly, your patch will only help with the list call in fsck.py, but you will still get errors with the others. Is that right?

If so, this will turn into a game of whack-a-mole as you keep adding additional translation rules.

What I'm wondering is therefore:  could we instead modify the calls to `backend.list()`, so that no backend transformation is needed?  I'm pretty sure that e.g. in fsck.s3ql, we could simply retrieve the full object list and skip over everything that does not start with `s3ql_data` on the client (the number of such objects is very small by comparison).

I am not so sure about the calls in database.py, but here we could conceivably change the object name (since there aren't that many of these objects they could be re-uploaded). However, I also don't understand why you are not currently running into trouble with them.

**Comment by DarkCaster** [comment]
> If I understood you correctly, your patch will only help with the list call in fsck.py, but you will still get errors with the others. Is that right?

Not quite like that. In brief - It allow any query in backend.list to work properly. All other utilities should also work flawlessly as if this is a normal s3c backend. For example, I already converted my FS revision with s3qladm.

Here is a more detailed explanation how storjs3.backend.list is working now:
1. for exact `s3ql_data_` queries, or for any queries started with partial occurrence of a string `s3ql_data_` :
1.1. it will transform this query into `s3ql_data/` query
1.2. query executed at s3/server-side and the results are fetched page by page
1.3. every key returned in `s3ql_data/<number>` form converted to `s3ql_data_<number>` form as if initial query was `s3ql_data_`
2. for non exact `s3ql_data_` queries:
2.1. it will query all objects from `s3ql_other` prefix with `s3ql_other/` query
2.2. query executed at s3/server side, ALL results are fetched before next step
2.3. perform converting resulted keys from `s3ql_other/<base64_encoded_original_key>` to `<original_key>` while iterating over the converted keys
2.4. perform manual client side filtering of back-converted keys over original query, appending the results to the result from p.1 if any.

**Comment by DarkCaster** [comment]
> If so, this will turn into a game of whack-a-mole as you keep adding additional translation rules.

It should work with any query. However, it can only perform server-side filtering without fetching all the results on "s3ql_data_" query. And we only talking about `backend.list` calls. For non "s3ql_data_" queries it will fetch the whole keys list from "s3ql_other/" prefix and perform the search manually, client-side. This scheme should work fine as long as we have a finite and relatively small amount of non "s3ql_data_" keys. That's the main flaw of it. But I assume that `backend.list` (with any search prefix) is called not that often.

On-the-fly translation of key-paths' for other calls should work without any limitations or significant performance loss at all.

**Comment by DarkCaster** [comment]
>> What I'm wondering is therefore: could we instead modify the calls to backend.list(), so that no backend transformation is needed? I'm pretty sure that e.g. in fsck.s3ql, we could simply retrieve the full object list and skip over everything that does not start with s3ql_data on the client (the number of such objects is very small by comparison).

Probably, yes. Technically, I already doing something like this, but skipping s3ql_data objects by server-side.

**Comment by DarkCaster** [comment]
> I am not so sure about the calls in database.py, but here we could conceivably change the object name (since there aren't that many of these objects they could be re-uploaded). However, I also don't understand why you are not currently running into trouble with them.

Sorry, I'm not quite understand what you mean. Every call committed via storjs3 backend clall should be silently translated and work as intended. That was my original idea - to make a wrapper and not to change any other s3 logic. If some code at s3ql codebase performs s3 call not via storjs3/s3 backend classes, but directly, that of course, will not work and can lead to the nasty FS corruption at some moment. Sorry, maybe I'm not fully understood what you mean. Can you rephrase, please ?

**Comment by Nikratio** [comment]
> > I am not so sure about the calls in database.py, but here we could conceivably change the object name (since there aren't that many of these objects they could be re-uploaded). However, I also don't understand why you are not currently running into trouble with them.
> 
> Sorry, I'm not quite understand what you mean. Every call committed via storjs3 backend class should be silently translated and work as intended. That was my original idea - to make a wrapper and not to change any other s3 logic. If some code at s3ql codebase performs s3 call not via storjs3/s3 backend classes, but directly, that of course, will not work and can lead to the nasty FS corruption at some moment. Sorry, maybe I'm not fully understood what you mean. Can you rephrase, please ?

I mean that, instead of creating a wrapper, we could change the s3ql logic so that there is no wrapper needed.

For example, if we renamed the `s3ql_params_<xx>` and `s3ql_metadata_<yy>` objects to `s3ql_params/<xx>` and `s3ql_metadata_<yy>`, the code would call list with a prefix ending in `/` and there would be no need for any of the renaming.

This is probably not worth the trouble just for one buggy backend, but I have similar doubts that we want to add the another backend for the sole purpose of working around a bug. Is there any prospect of this getting fixed on the StorJ side?

**Comment by DarkCaster** [comment]
> Is there any prospect of this getting fixed on the StorJ side?
Looks like no. As I understood their answer - object names are encrypted in-between forward slash delimiters. Arbitrary prefix listing/search will only work for buckets with small number of objects (< 100K), because it involve decryption of full object name. For buckets with more objects in it - listing will only work with prefix ended with `/`

**Comment by DarkCaster** [comment]
> I mean that, instead of creating a wrapper, we could change the s3ql logic so that there is no wrapper needed.

It would be great, but I'm definitely not up to the task. In addition, there are no guarantees that the Storj developers will not eventually break something else on their side that will have to be repaired separately from the main S3 code base.

The whole S3 thing seems cursed, in my experience. Every single vendor I've tried for my projects for the long years cannot deliver even basic S3 functionality to work identically to others. To be honest, I don’t quite understand how it’s even possible in such conditions to develop software that, in principle, shows some acceptable level of stability.

**Comment by Nikratio** [comment]
Well, in that case maybe the right question to ask is if you truly want to trust these vendors with your data?

The "task" is just some string replacements. But you have convinced me that this isn't the right approach.

Let's keep this issue open for a while and see if others are interested. If you are the only one who wants to use S3QL with StorJS, then it may make more sense if you keep using this patch locally (should be fairly easy especially since it doesn't change S3QL internals).

**Comment by DarkCaster** [comment]
> Well, in that case maybe the right question to ask is if you truly want to trust these vendors with your data?

This is the right question, indeed. For my use case, there are not many options left anyway. This particular problem was the first major issue I experienced with them for the last 2 years, because of which I had to stop my production for almost a week. That why I contacted you with possible solution. Nothing permanent exists nowadays anyway, I'll switch to other service someday. So far they haven’t lost my data at least, and I’ve had temporary service availability problems even with AWS.

> But you have convinced me that this isn't the right approach.

Yes, I was in a hurry to implement the minimalistic solution that just works and easy to maintain in a first place. Maybe at least this will help somebody else.
